contracts/btc-light-client/src/bitcoin.rs

```diff
@@ -1,54 +1,277 @@
+    header
+        .validate_pow(target)
+        .map_err(HeaderError::InvalidProofOfWork)?;
+
+    Ok(())
+}
+
+/// > Ensure the difficulty specified in the block header matches
+/// > the calculated difficulty based on the previous block and
+/// > difficulty retarget rules.
+///
+/// Note: While the naming mirrors the btcd implementation, this function only performs
+/// the difficulty adjustment check (not the full header context validation in btcd).
+///
+/// https://pkg.go.dev/github.com/btcsuite/btcd@v0.24.2/blockchain#CheckBlockHeaderContext
+fn check_block_header_context(
+    storage: &dyn Storage,
+    chain_params: &Params,
+    prev_block_height: u32,
+    prev_block_header: &BlockHeader,
+    header: &BlockHeader,
+) -> Result<(), HeaderError> {
+    let expected_target =
+        get_next_work_required(storage, prev_block_height, prev_block_header, chain_params)?;
+
+    let expected_bits = expected_target.to_compact_lossy().to_consensus();
+
+    let actual_target = header.target();
+    let actual_bits = actual_target.to_compact_lossy().to_consensus();
+
+    if actual_bits != expected_bits {
+        return Err(HeaderError::BadDifficultyBits {
+            got: actual_target,
+            expected: expected_target,
+        });
+    }
+
+    Ok(())
+}
+
+/// Usually, it's just the target of last block. However, if we are in a retarget period,
+/// it will be calculated from the last 2016 blocks (about two weeks for Bitcoin mainnet).
+///
+/// <https://github.com/bitcoin/bitcoin/blob/89b910711c004c21b7d67baa888073742f7f94f0/src/pow.cpp#L13>
+fn get_next_work_required(
+    storage: &dyn Storage,
+    last_block_height: u32,
+    last_block: &BlockHeader,
+    params: &Params,
+) -> Result<Target, HeaderError> {
+    if params.no_pow_retargeting {
+        return Ok(last_block.target());
+    }
+
+    let height = last_block_height + 1;
+
+    let difficulty_adjustment_interval = params.difficulty_adjustment_interval() as u32;
+
+    // Only change once per difficulty adjustment interval.
+    if height >= difficulty_adjustment_interval && height % difficulty_adjustment_interval == 0 {
+        let last_retarget_height = height - difficulty_adjustment_interval;
+
+        let retarget_header_info = get_header(storage, last_retarget_height)?;
+        let retarget_header: BlockHeader =
+            bitcoin::consensus::deserialize(&retarget_header_info.header)?;
+
+        let first_block_time = retarget_header.time;
+
+        // timestamp of last block
+        let last_block_time = last_block.time;
+
+        Ok(calculate_next_work_required(
+            Uint256::from_be_bytes(last_block.target().to_be_bytes()),
+            first_block_time.into(),
+            last_block_time.into(),
+            params,
+        ))
+    } else {
+        Ok(last_block.target())
+    }
+}
+
+// <https://github.com/bitcoin/bitcoin/blob/89b910711c004c21b7d67baa888073742f7f94f0/src/pow.cpp#L49-L72>
+fn calculate_next_work_required(
```

Comment:
maybe we can re-use functions from rust bitcoin - https://github.com/rust-bitcoin/rust-bitcoin/blob/f7274a57c36d0b8d8cc528426e269c27d496bd5f/bitcoin/src/pow.rs#L374 ?  They should be compatible with `cpp` and `golang` impls.

Reply:
If I remember correctly, it was a bit inconvenient to reuse the APIs from rust-bitcoin at the time. But that might not be the case anymore. I can take another look and see if things work smoothly now.